consumer/fileobservable.go

```diff
@@ -0,0 +1,226 @@
+// limitations under the License.
+
+package consumer
+
+import (
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
+	"github.com/trivago/tgo/tio"
+	"github.com/trivago/tgo/tsync"
+)
+
+type observableFile struct {
+	handle         *os.File
+	fileName       string
+	offsetFileName string
+	cursor         fileCursor
+	buffer         *tio.BufferedReader
+
+	lastStatCheck time.Time
+	retryDelay    time.Duration
+	pollDelay     time.Duration
+	log           logrus.FieldLogger
+}
+
+type fileCursor struct {
+	whence int
+	offset int64
+}
+
+func (fs *observableFile) close() error {
+	if fs.handle != nil {
+		return fs.handle.Close()
+	}
+	return nil
+}
+
+func (fs *observableFile) getActualFilename() string {
+	actualFileName := fs.fileName
+	if evalFileName, err := filepath.EvalSymlinks(actualFileName); err == nil {
+		actualFileName = evalFileName
+	}
+	if absFileName, err := filepath.Abs(actualFileName); err == nil {
+		actualFileName = absFileName
+	}
+	return actualFileName
+}
+
+func (fs *observableFile) hasRotated(currentName string) bool {
+	if fs.handle == nil {
+		return false
+	}
+
+	actualFileName := fs.getActualFilename()
+	if fs.handle.Name() != actualFileName {
+		return true
+	}
+
+	if time.Since(fs.lastStatCheck) < fs.pollDelay {
+		return false
+	}
+
+	newStat, newStatErr := os.Stat(actualFileName)
+	oldStat, oldStatErr := fs.handle.Stat()
+	fs.lastStatCheck = time.Now()
+	return newStatErr != oldStatErr || !os.SameFile(newStat, oldStat)
+}
+
+func (fs *observableFile) storeOffset() {
+	fs.cursor.offset, _ = fs.handle.Seek(0, io.SeekCurrent)
+	offsetAsString := strconv.FormatInt(fs.cursor.offset, 10)
+	if err := ioutil.WriteFile(fs.offsetFileName, []byte(offsetAsString), 0644); err != nil {
+		fs.log.WithError(err).Error("Failed to store offset")
+	}
+}
+
+func (fs *observableFile) scrape(fileName string, enqueue func([]byte), onRotate func()) {
+	// Try to open the current file
+	if fs.handle == nil {
+		handle, err := os.OpenFile(fileName, os.O_RDONLY, 0444)
+		if err != nil {
+			fs.log.Warning("Failed to open file")
+			time.Sleep(fs.retryDelay)
+			return // wait between retries
+		}
+
+		if fs.cursor.offset, err = handle.Seek(fs.cursor.offset, fs.cursor.whence); err != nil {
+			fs.log.WithError(err).Warning("Failed to seek to given offset")
+		}
+		fs.handle = handle
+	}
+
+	// Try to scrape the file
+	err := fs.buffer.ReadAll(fs.handle, enqueue)
+
+	switch err {
+	case nil:
+	case io.EOF:
+		if fs.hasRotated(fileName) {
+			fs.log.Info("File rotated")
+			fs.handle.Close()
+			fs.handle = nil
+			fs.buffer.Reset(0)
+
+			fs.cursor.whence = io.SeekStart
+			fs.cursor.offset = 0
+
+			onRotate()
+		}
+	default:
+		fs.log.WithError(err).Error("Failed to read file")
+		fs.handle.Close()
+		fs.handle = nil
+		fs.buffer.Reset(0)
+	}
+}
+
+func (fs *observableFile) observePoll(enqueue func([]byte), done chan struct{}) {
```

Comment:
It would be nice to use `done <-chan struct{}` instead, so one can easily see that the channel is only to be received from.